File: android/src/main/java/com/fl/rangers/app/log/FlRangersAppLogPlugin.java
package com.fl.rangers.app.log;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import com.bytedance.applog.AppLog;
import com.bytedance.applog.ILogger;
import com.bytedance.applog.InitConfig;
import com.bytedance.applog.UriConfig;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlRangersAppLogPlugin
 */
public class FlRangersAppLogPlugin implements FlutterPlugin, MethodCallHandler {

    private Context context;
    private MethodChannel channel;

    @Override
    public void onMethodCall(MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "initRangersAppLog":
                InitConfig initConfig = new InitConfig((String) Objects.requireNonNull(call.argument("appid")), (String) call.argument("channel"));
                initConfig.setAutoStart(true);
                initConfig.setAbEnable((Boolean) call.argument("enable_ab"));
                AppLog.setEncryptAndCompress((Boolean) call.argument("enable_encrypt"));
                if ((Boolean) call.argument("enable_log")) {
                    initConfig.setLogger(new ILogger() {
                        @Override
                        public void log(String s, Throwable throwable) {
                            Log.d("AppLog------->: ", "" + s);
                        }
                    });
                }
                if ((String) call.argument("host") != null) {
                    initConfig.setUriConfig(UriConfig.createByDomain((String) call.argument("host"), null));
                }
                AppLog.init(context, initConfig);
                break;
            case "getDeviceId":
                result.success(AppLog.getDid());
                break;
            case "getAbSdkVersion":
                result.success(AppLog.getAbSdkVersion());
                break;
            case "getABTestConfigValueForKey":
                result.success(AppLog.getAbConfig((String) call.argument("key"), call.argument("default")));
                break;
            case "onEventV3":
                String eventName = (String) call.argument("event");
                assert eventName != null;
                AppLog.onEventV3(eventName, getJsonFromMap(call, "param"));
                break;
            case "setUserUniqueId":
                AppLog.setUserUniqueID((String) call.argument("uuid"));
                break;
            case "setHeaderInfo":
                AppLog.setHeaderInfo((HashMap<String, Object>) call.argument("customHeader"));
                break;
            case "profileSet":
                AppLog.profileSet(getJsonFromMap(call, "profileDict"));
                break;
            case "profileSetOnce":
                AppLog.profileSetOnce(getJsonFromMap(call, "profileDict"));
                break;
            case "profileIncrement":
                AppLog.profileIncrement(getJsonFromMap(call, "profileDict"));
                break;
            case "profileAppend":
                AppLog.profileAppend(getJsonFromMap(call, "profileDict"));
                break;
            case "profileUnSet":
                AppLog.profileUnset((String) call.argument("key"));
                break;
            case "getAllAbTestConfig":
                break;
            case "removeHeaderInfo":
                AppLog.removeHeaderInfo((String) call.argument("key"));
            default:
                result.notImplemented();
                break;
        }
    }

    private JSONObject getJsonFromMap(MethodCall call, String param) {
        HashMap<String, Object> paramMap = (HashMap<String, Object>) call.argument(param);
        JSONObject paramJson = new JSONObject();
        assert paramMap != null;
        for (Map.Entry<String, Object> entry : paramMap.entrySet()) {
            try {
                paramJson.put(entry.getKey(), entry.getValue());
            } catch (JSONException ignored) {
            }
        }
        return paramJson;
    }


    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "com.fl.rangers.app.log");
        context = binding.getApplicationContext();
        channel.setMethodCallHandler(this);
        FlRangersAppLogEvent.init(binding);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

}

File: android/src/main/kotlin/com/fl/rangers/app/log/FlRangersAppLogPlugin.kt
package com.fl.rangers.app.log

import android.content.Context
import android.util.Log
import com.bytedance.applog.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import org.json.JSONObject

/**
 * FlRangersAppLogPlugin
 */
class FlRangersAppLogPlugin : FlutterPlugin, MethodCallHandler {
    private var context: Context? = null
    private var channel: MethodChannel? = null

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.fl.rangers.app.log")
        context = binding.applicationContext
        channel!!.setMethodCallHandler(this)
        val dataObserver = InnerDataObserver(channel!!)
        AppLog.addDataObserver(dataObserver)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        channel!!.setMethodCallHandler(null)
        AppLog.removeAllDataObserver()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initRangersAppLog" -> {
                val initConfig = InitConfig(
                    call.argument("appId")!!, call.argument("channel")!!
                )
                initConfig.setAutoStart(true)
                initConfig.isAbEnable = call.argument("enable_ab")!!
                AppLog.setEncryptAndCompress(call.argument("enable_encrypt")!!)
                if (call.argument<Boolean>("enable_log")!!) {
                    initConfig.logger = ILogger { s, _ -> Log.d("AppLog------->: ", "" + s) }
                }
                val host = call.argument<String?>("host")
                if (host != null) {
                    initConfig.uriConfig = UriConfig.createByDomain(host, null)
                }
                AppLog.init(context!!, initConfig)
                result.success(true)
            }
            "getDeviceId" -> result.success(AppLog.getDid())
            "getAbSdkVersion" -> result.success(AppLog.getAbSdkVersion())
            "getAllAbTestConfig" -> result.success(null)
            "getABTestConfigValueForKey" -> result.success(
                AppLog.getAbConfig(
                    call.argument("key"), call.argument("default")
                )
            )
            "onEventV3" -> {
                AppLog.onEventV3(call.argument<String>("event")!!, getJsonFromMap(call, "param"))
                result.success(true)
            }
            "setUserUniqueId" -> {
                AppLog.setUserUniqueID(call.arguments as String)
                result.success(true)
            }
            "setHeaderInfo" -> {
                AppLog.setHeaderInfo(call.arguments as HashMap<String, Any>?)
                result.success(true)
            }
            "removeHeaderInfo" -> {
                AppLog.removeHeaderInfo(call.arguments as String)
                result.success(true)
            }
            "profileSet" -> {
                AppLog.profileSet(getJsonFromMap(call, null))
                result.success(true)
            }
            "profileSetOnce" -> {
                AppLog.profileSetOnce(getJsonFromMap(call, null))
                result.success(true)
            }
            "profileIncrement" -> {
                AppLog.profileIncrement(getJsonFromMap(call, null))
                result.success(true)
            }
            "profileAppend" -> {
                AppLog.profileAppend(getJsonFromMap(call, null))
                result.success(true)
            }
            "profileUnSet" -> {
                AppLog.profileUnset(call.arguments as String)
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    private fun getJsonFromMap(call: MethodCall, paramKey: String?): JSONObject {
        var paramMap = call.arguments as Map<*, *>
        if (paramKey != null) paramMap = paramMap[paramKey] as Map<*, *>
        val paramJson = JSONObject()
        for ((key, value) in paramMap) {
            paramJson.put(key as String, value)
        }
        return paramJson
    }


    private class InnerDataObserver(private val channel: MethodChannel) : IDataObserver {
        override fun onIdLoaded(s: String, s1: String, s2: String) {}
        override fun onRemoteIdGet(
            b: Boolean, s: String, s1: String, s2: String, s3: String, s4: String, s5: String
        ) {
        }

        override fun onRemoteConfigGet(b: Boolean, jsonObject: JSONObject) {}
        override fun onRemoteAbConfigGet(b: Boolean, jsonObject: JSONObject) {
            channel.invokeMethod("onABTest", jsonObject.toString())
        }

        override fun onAbVidsChange(s: String, s1: String) {
            channel.invokeMethod("onABTestVidsChanged", arrayListOf(s, s1));
        }
    }
}